README.md

```diff
@@ -0,0 +1,274 @@
+Pynisher is a library to limit resources during the running of synchronous functions.
+
+## Quick Overview
+
+Limit the time a process can take
+```python
+import pynisher
+
+
+def sleepy(x: int) -> int:
+    time.sleep(x)
+    return x
+
+# You can also use `cpu_time` instead
+with pynisher.limit(sleepy, wall_time=7) as limited_sleep:
+    x = limited_sleep(10)  # Will raise a TimeoutException
+```
+
+Limit the memory usage in a sequential manner
+```python
+from pynisher import limit, MemoryLimitException, WallTimeoutException
+
+
+def train_memory_hungry_model(X, y) -> Model:
+    # ... do some thing
+    return Model
+
+model_trainer = limit(
+    train_memory_hungry_model,
+    name="Name for the process",
+    memory=(500, "MB"),
+    wall_time=(1.5, "h")  # 1h30m
+)
+
+try:
+    model = model_trainer(X, y)
+except (WallTimeoutException, MemoryLimitException):
+    model = None
+```
+
+Passing `raises=False` means it will hide all errors and will return `EMPTY` if
+there is no result to give back.
+
+```python
+from pynisher import limit, EMPTY
+
+def f():
+    raise ValueError()
+
+limited_f = limit(f, wall_time=(2, "m"), raises=False)
+result = limited_f()
+
+if result is not EMPTY:
+    # ...
+```
+
+
+You can even use the decorator, in which case it will always be limited.
+Please note in [Details](#details) that support for this is limited and mostly
+for Linux.
+```python
+from pynisher import restricted
+
+@restricted(wall_time=1, raises=False)
+def notify_remote_server() -> Response:
+    """We don't care that this fails, just give it a second to try"""
+    server = block_until_access(...)
+    response = server.notify()
+
+notify_remote_server()
+# ... continue on even if it failed
+```
+
+You can safely raise errors from inside your function and the same kind of error will be reraised
+with a traceback.
+```python
+from pynisher import limit
+
+
+def f():
+    raise MyCustomException()
+
+limited_f = limit(f)
+
+try:
+    limited_f()
+except MyCustomException as e:
```

Comment:
Yes it will, the code version of the traceback is no retrievable or picklable so all we can do is send back a str representation of it.

Reply:
This sounds very dangerous to me. Let's assume the user imports pytorch only in the pynished function. Then, pytorch would be passed to the parent process and populate memory there. Similar, for auto-sklearn, the different model libraries in scikit-learn that are only imported in the subprocesses (SVM, GBDT, etc) would be passed to the main process and populate memory there. Therefore, it would be great if one could disable the exception being passed to the main process. In that case, In my opinion, only the string representation of the exception should be presented to the user.

Reply:
I agree in principle but I think it's worth clarifying on your response with respect to pickling objects through pipes.To me it would make sense to make it an option that defaults to pass the same error through.

More specficailly, only a short reference to `pytorch` is pickled and transferred through the pipe not the entire library as bytes. This reference is very small and is jsut a string representation of the module name to import. There is also the case that for a user to define the function, they need to import `pytorch` in the master process anywho:

```python
import pytorch

def f():
    # do pytorch things
    # raise PytorchError

limited_func = limit(f)
result = limited_func()  # Pytorch already exists in the master process.
```

The exception to this is of course imports that are scoped
```python
def f():
    import pytorch
    # do pytorch things
    # raise PytorchError

limited_func = limit(f)
limited_func()  # Pytorch now imported into the main process
```

However, if you're working with something like `pytorch`, you likely have it in the master process somewhere. The first use case seems like the most common and hence why I would make it an option that defaults to raise the same error class as the one in the subprocess